### DISCUSSION TITLE
Note_Off is displayed as Note_On

### DISCUSSION THREAD

**Discussion by sharmilamani** [mido/mido#605] [Q&A] [answered]
Hi,
How to identify the start time and duration of note played from mid file ?
as i went through the mido read me time in message refers to wait time from previous event. is this the start time of the note ? also i don't see a "note_off message".

Below is a midi file printed using mido. In this file, the key c4 is pressed and released 10 times.
Kindly help in understanding
1) start of each key press in seconds ?
2) time duration / end of each key press ?
3) time gap between two key press ( start of second key - end of first key)

MidiFile(type=1, ticks_per_beat=24, tracks=[
  MidiTrack([
    MetaMessage('set_tempo', tempo=500000, time=0),
    MetaMessage('time_signature', numerator=4, denominator=4, clocks_per_click=24, notated_32nd_notes_per_beat=8, time=0),
    MetaMessage('key_signature', key='C', time=0),
    MetaMessage('end_of_track', time=0)]),
  MidiTrack([
    Message('program_change', channel=0, program=4, time=0),
    Message('note_on', channel=0, note=60, velocity=46, time=54),
    Message('note_on', channel=0, note=60, velocity=0, time=30),
    Message('note_on', channel=0, note=60, velocity=40, time=30),
    Message('note_on', channel=0, note=60, velocity=0, time=36),
    Message('note_on', channel=0, note=60, velocity=33, time=36),
    Message('note_on', channel=0, note=60, velocity=0, time=36),
    Message('note_on', channel=0, note=60, velocity=38, time=36),
    Message('note_on', channel=0, note=60, velocity=0, time=36),
    Message('note_on', channel=0, note=60, velocity=34, time=36),
    Message('note_on', channel=0, note=60, velocity=0, time=42),
    Message('note_on', channel=0, note=60, velocity=40, time=36),
    Message('note_on', channel=0, note=60, velocity=0, time=36),
    Message('note_on', channel=0, note=60, velocity=45, time=36),
    Message('note_on', channel=0, note=60, velocity=0, time=42),
    Message('note_on', channel=0, note=60, velocity=48, time=48),
    Message('note_on', channel=0, note=60, velocity=0, time=42),
    Message('note_on', channel=0, note=60, velocity=51, time=36),
    Message('note_on', channel=0, note=60, velocity=0, time=42),
    Message('note_on', channel=0, note=60, velocity=28, time=24),
    Message('note_on', channel=0, note=60, velocity=0, time=42),
    MetaMessage('end_of_track', time=0)])

**Comment by jnguyen1098** [comment, marked as answer]
According to the [MIDI 1.0 specification](https://midi.org/midi-1-0-core-specifications), page 73, a MIDI note-on event (`9n`) with a velocity of 0 is considered a note-off. In your example, every `note_on` event is immediately followed with another `note_on` but with a `velocity=0`, thereby functioning as the `note_off`(s) you look for.

The advantage of this is that, assuming an uninterrupted stream of note-on and note-off events, a space-conscious implementation may take advantage of the running status. Your file likely employed such encoding methods.

I haven't looked at the mido codebase, so my answer is mostly related to MIDI as a whole, not this project in particular, which appears to be completely in-spec.

Edit: related discussions:

- https://github.com/mido/mido/issues/130
- https://github.com/mido/mido/issues/158
- https://www.kvraudio.com/forum/viewtopic.php?t=291892

**Reply by rdoursenaud** [reply]
@jnguyen1098 perfectly explained the reason. It is not a bug but an intrinsic design of the MIDI protocol.
